File: src/engine/script/runtime/tests/frame_navigation/file_reader.rs
use super::*;

#[test]
fn file_reader_encoding_uses_label_then_mime_with_bom_precedence() {
    let (dom, mut runtime) = start(
        r#"<body><script>
        window.results=[];
        const read=(bytes,type,label)=>{const r=new FileReader();r.onload=()=>results.push(r.result);r.readAsText(new Blob([new Uint8Array(bytes)],{type}),label)};
        read([0x80],'text/plain;charset=windows-1252','invalid-label');
        read([0xe9],'text/plain;charset=utf-8','iso-8859-1');
        read([0xff,0xfe,0x41,0],'text/plain','utf-8');
        read([0xc3,0xa9],'text/plain','invalid-label');
    </script>"#,
    );
    drain(&mut runtime);
    evaluate(
        &mut runtime,
        &dom,
        "if(results.join(',')!=='€,é,A,é') throw Error(results);",
    );
}

#[test]
fn file_reader_modes_preserve_private_bytes_and_async_state() {
    let (dom, mut runtime) = start(
        r#"<body><script>
        window.results=[];window.events=[];
        const blob=new Blob(['hello'],{type:'text/plain'});
        blob.arrayBuffer=()=>{throw Error('author method')};
        Object.defineProperty(blob,'type',{get(){throw Error('author getter')}});
        const methods=['readAsText','readAsBinaryString','readAsDataURL','readAsArrayBuffer'];
        for(const method of methods) {
            const r=new FileReader();
            if(r.readyState!==r.EMPTY || r.result!==null || r.error!==null) throw Error('initial state');
            r.onloadstart=e=>events.push(e.type+':'+e.isTrusted);
            r.onload=e=>{
                if(r.readyState!==r.DONE || r.error!==null || !e.lengthComputable || e.total!==5) throw Error('completion');
                results.push(method+':'+(r.result instanceof ArrayBuffer ? new TextDecoder().decode(r.result) : r.result));
            };
            r[method](blob);
            if(r.readyState!==r.LOADING || results.length) throw Error('synchronous read');
            try {r.readAsText(blob);throw Error('accepted concurrent read')}
            catch(e){if(e.name!=='InvalidStateError') throw e}
        }
    </script>"#,
    );
    drain(&mut runtime);
    evaluate(
        &mut runtime,
        &dom,
        r#"
        if(results.sort().join('|')!=='readAsArrayBuffer:hello|readAsBinaryString:hello|readAsDataURL:data:text/plain;base64,aGVsbG8=|readAsText:hello')
            throw Error(JSON.stringify(results));
        if(events.length!==4 || events.some(e=>e!=='loadstart:true')) throw Error(events);
        if('__fileReaderSnapshot' in globalThis) throw Error('private hook');
    "#,
    );
}

#[test]
fn file_reader_abort_cancels_old_tasks_and_read_chaining_omits_old_loadend() {
    let (dom, mut runtime) = start(
        r#"<body><script>
        window.events=[];const r=new FileReader(),blob=new Blob(['a']);
        r.onloadstart=()=>events.push('start');
        r.onabort=()=>events.push('abort');
        r.onloadend=()=>events.push('end');
        r.onload=()=>{events.push('load');if(events.filter(e=>e==='load').length===1)r.readAsText(blob)};
        r.readAsText(blob);r.abort();
        if(r.readyState!==r.DONE || r.result!==null || r.error!==null) throw Error('abort state');
        r.readAsText(blob);
    </script>"#,
    );
    drain(&mut runtime);
    evaluate(
        &mut runtime,
        &dom,
        "if(events.join(',')!=='abort,end,start,load,start,load,end') throw Error(events);",
    );
}

File: src/engine/script/bootstrap/file_reader.js
// File API read operation and abort algorithms: https://w3c.github.io/FileAPI/
(() => {
    'use strict';
    const [snapshot, concatBytes, bytesToBase64] = globalThis.__fileReaderSnapshot;
    delete globalThis.__fileReaderSnapshot;
    const schedule = globalThis.setTimeout, cancel = globalThis.clearTimeout;
    const trusted = globalThis.__markTrustedEvent, Progress = globalThis.ProgressEvent;
    const host = globalThis.__hostCall, Exception = globalThis.DOMException;
    const dispatch = EventTarget.prototype.dispatchEvent;
    const add = EventTarget.prototype.addEventListener, remove = EventTarget.prototype.removeEventListener;
    const states = new WeakMap();
    const stateOf = reader => {
        const state = states.get(reader);
        if (!state) throw new TypeError('Invalid FileReader receiver');
        return state;
    };
    const string = value => {
        if (typeof value === 'symbol') throw new TypeError('Encoding must be a DOMString');
        return String(value);
    };
    const decode = (bytes, type, label) => {
        const charset = /(?:^|;)\s*charset\s*=\s*(?:"([^"]*)"|([^;\s]*))/i.exec(type);
        return host('fileReadText', bytes, label ?? '', charset ? charset[1] ?? charset[2] : '');
    };
    const packageData = (bytes, mode, type, label) => {
        if (mode === 'buffer') return bytes.buffer;
        if (mode === 'text') return decode(bytes, type, label);
        if (mode === 'url') return 'data:' + (type || 'application/octet-stream') + ';base64,' + bytesToBase64(bytes);
        let result = '';
        for (let offset = 0; offset < bytes.length; offset += 0x4000)
            result += String.fromCharCode(...bytes.subarray(offset, offset + 0x4000));
        return result;
    };
    const fire = (reader, type, loaded, total) => dispatch.call(reader, trusted(new Progress(type, {
        lengthComputable: true, loaded, total
    })));
    const read = (reader, blob, mode, label) => {
        const state = stateOf(reader), data = snapshot(blob);
        if (label !== undefined) label = string(label);
        if (state.readyState === 1) throw new Exception('A read is already in progress', 'InvalidStateError');
        state.readyState = 1; state.result = state.error = null;
        state.loaded = 0; state.total = data.size;
        const generation = ++state.generation;
        const current = () => generation === state.generation && state.readyState === 1;
        state.task = schedule(() => {
            if (!current()) return;
            fire(reader, 'loadstart', 0, data.size);
            if (!current()) return;
            state.task = schedule(() => {
                if (!current()) return;
                let result, error;
                try { result = packageData(concatBytes(data.chunks), mode, data.type, label); }
                catch (_) { error = new Exception('Could not read Blob bytes', 'NotReadableError'); }
                if (!error) {
                    state.loaded = data.size;
                    fire(reader, 'progress', data.size, data.size);
                    if (!current()) return;
                }
                state.readyState = 2; state.task = null;
                state.result = error ? null : result; state.error = error || null;
                fire(reader, error ? 'error' : 'load', state.loaded, data.size);
                // A load/error handler may start a new read. Its state must not be
                // overwritten, nor may the previous operation emit loadend.
                if (state.readyState !== 1) fire(reader, 'loadend', state.loaded, data.size);
            }, 0);
        }, 0);
    };
    class FileReader extends EventTarget {
        constructor() {
            super();
            states.set(this, {readyState:0, result:null, error:null, generation:0,
                task:null, loaded:0, total:0, handlers:new Map()});
        }
        get readyState() { return stateOf(this).readyState; }
        get result() { return stateOf(this).result; }
        get error() { return stateOf(this).error; }
        readAsArrayBuffer(blob) { read(this, blob, 'buffer'); }
        readAsText(blob, encoding) { read(this, blob, 'text', encoding); }
        readAsDataURL(blob) { read(this, blob, 'url'); }
        readAsBinaryString(blob) { read(this, blob, 'binary'); }
        abort() {
            const state = stateOf(this);
            state.result = null;
            if (state.readyState !== 1) return;
            ++state.generation;
            cancel(state.task); state.task = null; state.readyState = 2;
            fire(this, 'abort', state.loaded, state.total);
            if (state.readyState !== 1) fire(this, 'loadend', state.loaded, state.total);
        }
    }
    for (const [name, value] of [['EMPTY',0], ['LOADING',1], ['DONE',2]])
        for (const target of [FileReader, FileReader.prototype])
            Object.defineProperty(target, name, {enumerable:true, value});
    for (const type of ['loadstart', 'progress', 'load', 'error', 'abort', 'loadend'])
        Object.defineProperty(FileReader.prototype, 'on' + type, {
            configurable:true, enumerable:true,
            get() { return stateOf(this).handlers.get(type)?.callback ?? null; },
            set(value) {
                const handlers = stateOf(this).handlers, previous = handlers.get(type);
                const callback = typeof value === 'function' ? value : null;
                if (previous && callback) { previous.callback = callback; return; }
                if (previous) { remove.call(this, type, previous.listener); handlers.delete(type); }
                if (callback) {
                    const item = {callback, listener:event => item.callback.call(this, event)};
                    handlers.set(type, item); add.call(this, type, item.listener);
                }
            }
        });
    Object.defineProperty(FileReader.prototype, Symbol.toStringTag, {configurable:true, value:'FileReader'});
    globalThis.FileReader = FileReader;
})();
